pkg/operator/status.go

```diff
@@ -150,65 +117,83 @@ func (optr *Operator) syncProgressingStatus() error {
 		if optr.inClusterBringup {
-			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeNormal, "OperatorVersionChanged", fmt.Sprintf("clusteroperator/machine-config-operator is bootstrapping to %v", optr.vStore.GetAll()))
-			coStatus.Message = fmt.Sprintf("Cluster is bootstrapping %s", optrVersion)
-			coStatus.Status = configv1.ConditionTrue
+			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeNormal, "OperatorVersionChanged", fmt.Sprintf("clusteroperator/machine-config is bootstrapping to %v", optr.vStore.GetAll()))
+			coStatusCondition.Message = fmt.Sprintf("Cluster is bootstrapping %s", optrVersion)
+			coStatusCondition.Status = configv1.ConditionTrue
 		}
 	} else {
 		// we can still be progressing during a sync (e.g. wait for master pool sync)
 		// but we want to fire the event only once when we're actually setting progressing and we
 		// weren't progressing before.
 		if !cov1helpers.IsStatusConditionTrue(co.Status.Conditions, configv1.OperatorProgressing) {
-			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeNormal, "OperatorVersionChanged", fmt.Sprintf("clusteroperator/machine-config-operator started a version change from %v to %v", co.Status.Versions, optr.vStore.GetAll()))
+			optr.eventRecorder.Eventf(mcoObjectRef, corev1.EventTypeNormal, "OperatorVersionChanged", fmt.Sprintf("clusteroperator/machine-config started a version change from %v to %v", co.Status.Versions, optr.vStore.GetAll()))
 		}
-		coStatus.Message = fmt.Sprintf("Working towards %s", optrVersion)
-		coStatus.Status = configv1.ConditionTrue
+		coStatusCondition.Message = fmt.Sprintf("Working towards %s", optrVersion)
+		coStatusCondition.Status = configv1.ConditionTrue
 	}
 
-	return optr.updateStatus(co, coStatus)
+	cov1helpers.SetStatusCondition(&co.Status.Conditions, coStatusCondition)
 }
 
-func (optr *Operator) updateStatus(co *configv1.ClusterOperator, status configv1.ClusterOperatorStatusCondition) error {
-	cov1helpers.SetStatusCondition(&co.Status.Conditions, status)
-	optr.setOperatorStatusExtension(&co.Status, nil)
-	_, err := optr.configClient.ConfigV1().ClusterOperators().UpdateStatus(context.TODO(), co, metav1.UpdateOptions{})
-	return err
+// This function updates the Cluster Operator's status via an API call only if there is an actual
+// to the status. Returns the final ClusterOperator object and an error if any.
+func (optr *Operator) updateClusterOperatorStatus(co *configv1.ClusterOperator, statusUpdate *configv1.ClusterOperatorStatus, syncErr error) (*configv1.ClusterOperator, error) {
+	// Update the operator status extension. This fetchs all the MCP status, and appends a syncerror if any.
+	optr.setOperatorStatusExtension(statusUpdate, syncErr)
+
+	newCO := co.DeepCopy()
+	newCO.Status = *statusUpdate
+
+	if !reflect.DeepEqual(co.Status, newCO.Status) {
+		// Attempt first update with the object passed in as it is likely to be the latest.
+		klog.Infof("CO Status Update, old resource version %s", co.ResourceVersion)
```

Comment:
Looking at https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4321/pull-ci-openshift-machine-config-operator-master-e2e-aws-ovn/1787473861281845248/artifacts/e2e-aws-ovn/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-operator-6b6f79868f-vcnm4_machine-config-operator.log there's a surprising amount of churn early on before the cluster settles and no updates happen at all, which I guess is expected

Reply:
Yeah, basically we get stuck in `syncRequiredPools` for a while, until all required pools are happy. A few parts of syncRequiredPools update the CO directly causing the flip/flops during the install. I decided to leave these direct updates in as the operator could potentially be stuck in this `syncRequiredPools` for a while and we'd want updates as the pool progresses through the install. 

I'd love to refactor `syncRequiredPools` someday and drop it down to just one update call though...